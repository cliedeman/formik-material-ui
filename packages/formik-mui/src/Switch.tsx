import * as React from 'react';
import {
  type SwitchProps as MuiSwitchProps,
  Switch as MuiSwitch, 
} from '@mui/material';
import type { FieldProps } from 'formik';
import invariant from 'tiny-warning';

export interface SwitchProps
  extends FieldProps,
    Omit<
      MuiSwitchProps,
      | 'checked'
      | 'name'
      | 'value'
      | 'defaultChecked'
      | 'form'
      // Excluded for conflict with Field type
      | 'type'
    > {
  type?: string;
}

export function fieldToSwitch({
  disabled,
  field: { onBlur: fieldOnBlur, ...field },
  form: { isSubmitting },
  type,
  onBlur,
  ...props
}: SwitchProps): MuiSwitchProps {
  if (process.env.NODE_ENV !== 'production') {
    invariant(
      type === 'checkbox',
      `property type=checkbox is missing from field ${field.name}, this can caused unexpected behaviour`
    );
  }

  return {
    disabled: disabled ?? isSubmitting,
    onBlur:
      onBlur ??
      function (e) {
        fieldOnBlur(e ?? field.name);
      },
    ...field,
    ...props,
  };
}

export function Switch(props: SwitchProps) {
  return <MuiSwitch {...fieldToSwitch(props)} />;
}

Switch.displayName = 'FormikMaterialUISwitch';
