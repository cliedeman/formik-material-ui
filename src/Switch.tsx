// @ts-ignore
import * as React from 'react';
import MuiSwitch, {
  SwitchProps as MuiSwitchProps,
} from '@material-ui/core/Switch';
import { FieldProps } from 'formik';
import { Omit } from './types';

export interface SwitchProps extends FieldProps, Omit<MuiSwitchProps, 'form'> {}

const Switch: React.ComponentType<SwitchProps> = ({
  field,
  form: { isSubmitting },
  disabled = false,
  ...props
}) => (
  <MuiSwitch
    disabled={isSubmitting || disabled}
    {...props}
    {...field}
    // TODO: is this the correct way?
    value={field.value ? 'checked' : ''}
  />
);

Switch.displayName = 'FormikMaterialUISwitch';

export default Switch;
