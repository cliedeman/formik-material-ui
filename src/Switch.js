/* @flow */
import Switch from '@material-ui/core/Switch';

import createComponent from './createComponent';

export default createComponent(
  Switch,
  ({ field, form: { isSubmitting }, disabled = false, ...props }) => ({
    disabled: isSubmitting || disabled,
    ...props,
    ...field,
    // TODO: is this the correct way?
    value: field.value ? 'checked' : '',
  })
);
