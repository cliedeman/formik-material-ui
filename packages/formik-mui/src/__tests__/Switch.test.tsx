import * as React from 'react';
import { test, expect } from 'vitest';
import { Formik, Form, Field } from 'formik';
import { render } from '@testing-library/react';

import { Switch } from '../Switch';

test('Switch Renders Correctly', () => {
  const { asFragment } = render(
    <Formik initialValues={{}} onSubmit={() => {}}>
      <Form>
        <Field component={Switch} name="test" type="checkbox" />
      </Form>
    </Formik>
  );

  expect(asFragment()).toMatchSnapshot();
});
