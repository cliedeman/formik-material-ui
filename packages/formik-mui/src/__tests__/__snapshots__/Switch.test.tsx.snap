// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Switch Renders Correctly 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <span
      class="MuiSwitch-root MuiSwitch-sizeMedium css-julti5-MuiSwitch-root"
    >
      <span
        class="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary css-byenzh-MuiButtonBase-root-MuiSwitch-switchBase"
      >
        <input
          class="PrivateSwitchBase-input MuiSwitch-input css-1m9pwf3"
          name="test"
          type="checkbox"
        />
        <span
          class="MuiSwitch-thumb css-jsexje-MuiSwitch-thumb"
        />
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </span>
      <span
        class="MuiSwitch-track css-1yjjitx-MuiSwitch-track"
      />
    </span>
  </form>
</DocumentFragment>
`;
